python/paddle/distributed/passes/auto_parallel_quantization.py

```diff
@@ -0,0 +1,255 @@
+
+        # 2. Prepare inputs
+        transform_pass_ops = []
+        quant_dequant_ops = []
+        quantize_op_types = [
+            'conv2d', 'depthwise_conv2d', 'mul', 'matmul', 'matmul_v2'
+        ]
+        for op_type in quantize_op_types:
+            if op_type in TRANSFORM_PASS_OP_TYPES:
+                transform_pass_ops.append(op_type)
+            elif op_type in QUANT_DEQUANT_PASS_OP_TYPES:
+                quant_dequant_ops.append(op_type)
+
+        weight_quantize_type = "channel_wise_abs_max" if self.get_attr(
+            'channel_wise_abs_max') else "abs_max"
+
+        # 3. Add quant op for ops which have parameters
+        transform_pass = QuantizationTransformPassV2(
+            scope=scope,
+            place=place,
+            weight_bits=self.get_attr('weight_bits'),
+            activation_bits=self.get_attr('activation_bits'),
+            skip_pattern=self.get_attr('not_quant_pattern'),
+            activation_quantize_type="moving_average_abs_max",
+            quantizable_op_type=transform_pass_ops,
+            weight_quantize_type=weight_quantize_type,
+            weight_quantize_func=None,
+            act_quantize_func=None,
+            weight_preprocess_func=None,
+            act_preprocess_func=None,
+            optimizer_func=None,
+            executor=None)
+        transform_pass.apply(main_graph)
+
+        # 4. Add quant op for ops which don't have parameter
+        quant_dequant_pass = AddQuantDequantPassV2(
+            scope=scope,
+            place=place,
+            quant_bits=self.get_attr('activation_bits'),
+            skip_pattern=self.get_attr('not_quant_pattern'),
+            quantizable_op_type=quant_dequant_ops)
+        quant_dequant_pass.apply(main_graph)
+
+        # 5. Gather quantitative information for the output
+        out_scale_training_pass = OutScaleForTrainingPass(scope=scope,
+                                                          place=place)
+        out_scale_training_pass.apply(main_graph)
+
+        # 6. Convert Graph back to Program
+        quant_program = main_graph.to_program()
+
+        # 7. get new prams_grads from quant_program
+        new_params_grads = []
+        for param, grad in params_grads:
+            if param.name not in quant_program.global_block().vars:
+                continue
+
+            new_param = quant_program.global_block().vars[param.name]
+            new_grad = quant_program.global_block().vars[grad.name]
+            new_params_grads.append((new_param, new_grad))
+
+        # 8. complete distributed attribution
+        # NOTE: hack implement, upgrading soon
+        for ib, block in enumerate(quant_program.blocks):
+            # recover origin ops' dist_attr and set quant ops' dist_attr
+            qat_offset = 0
+            for ip, quant_op in enumerate(block.ops):
+                quant_op_dist_attr = OperatorDistributedAttribute()
+
+                if "quantize" in quant_op.type or \
```

Comment:
shoud add kOpDeviceAttrName attr for quantization related op.  

and make sure the value for kOpDeviceAttrName of quantization related op is the same as the the corresponding backward op with same parameter.

Reply:
Done.